scripts/west_commands/runners/nrfutil.py

```diff
@@ -0,0 +1,117 @@
+# Copyright (c) 2023 Nordic Semiconductor ASA.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+'''Runner for flashing with nrfutil.'''
+
+import json
+import os
+import sys
+import subprocess
+import tempfile
+
+from runners.nrf_common import NrfBinaryRunner
+
+class NrfUtilBinaryRunner(NrfBinaryRunner):
+    '''Runner front-end for nrfutil.'''
+
+    def __init__(self, cfg, family, softreset, dev_id, erase=False,
+                 tool_opt=[], force=False, recover=False):
+
+        super().__init__(cfg, family, softreset, dev_id, erase,
+                         tool_opt, force, recover)
+        self._ops = []
+        self._op_id = 1
+
+    @classmethod
+    def name(cls):
+        return 'nrfutil'
+
+    @classmethod
+    def tool_opt_help(cls) -> str:
+        return 'Additional options for nrfutil, e.g. "--log-level"'
+
+    @classmethod
+    def do_create(cls, cfg, args):
+        return NrfUtilBinaryRunner(cfg, args.nrf_family, args.softreset,
+                                   args.dev_id, erase=args.erase,
+                                   tool_opt=args.tool_opt, force=args.force,
+                                   recover=args.recover)
+
+    def _exec(self, args):
+        try:
+            out = self.check_output(['nrfutil', '--json', 'device'] + args)
+        except subprocess.CalledProcessError as e:
+            # https://docs.python.org/3/reference/compound_stmts.html#except-clause
+            cpe = e
+            out = cpe.stdout
+        else:
+            cpe = None
+        finally:
+            # https://github.com/ndjson/ndjson-spec
+            out = [json.loads(s) for s in
+                   out.decode(sys.getdefaultencoding()).split('\n') if len(s)]
+            self.logger.debug(f'output: {out}')
+            if cpe:
+                if 'execute-batch' in args:
+                    for o in out:
+                        if o['type'] == 'batch_end' and o['data']['error']:
+                            cpe.returncode = o['data']['error']['code']
+                raise cpe
+
+        return out
+
+    def do_get_boards(self):
+        out = self._exec(['list'])
+        devs = []
+        for o in out:
+            if o['type'] == 'task_end':
+                devs = o['data']['data']['devices']
+        snrs = [dev['serialNumber'] for dev in devs]
+
+        self.logger.debug(f'Found boards: {snrs}')
+        return snrs
+
+    def do_require(self):
+        self.require('nrfutil')
+
+    def _insert_op(self, op):
+        op['operationId'] = f'{self._op_id}'
+        self._op_id += 1
+        self._ops.append(op)
+
+    def _exec_batch(self):
+        # prepare the dictionary and convert to JSON
+        batch = json.dumps({'family': f'{self.family}',
+                            'operations': [op for op in self._ops]},
+                            indent=4) + '\n'
+        fd, name = tempfile.mkstemp(prefix='nrfutil-batch-',
```

Comment:
This leaves behind the temporary file without cleaning it up -- see an alternative using tempfle.TemporaryDirectory as a context manager in jlink.py for a solution that cleans up after itself

Reply:
See my comment above to Gerard. I want the file to stay behind, because the whole logic in `nrfutil` is contained in this JSON file. So, in order to find out what `west flash` actually did is to look at the JSON file. The altenative would be to place the temporary file in the `build/` folder or naming it always the same to avoid leaving a bunch of files, what do you think?

Reply:
> The altenative would be to place the temporary file in the build/ folder

That sounds clean to me